### ISSUE TITLE
[sanitize] segfault in sanitize code may need a memory barrier.

### ISSUE THREAD

**Issue by garretrieger** [harfbuzz/harfbuzz#4478]
In a usage of harfbuzz in another library we're encountering a segfault when running a fuzzer. Happens here:

https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-layout-common.hh#L3911

I've checked and version.sanitize(c) is returning false as expected. However, the memory access to version.major is still happening which triggers the segfault. I suspect this is a memory barrier issue like: https://github.com/harfbuzz/harfbuzz/commit/f3151b6582a38fb1a377eb6070b8cecedb2ea711

**Comment by behdad** [comment]
Unfortunately this is a major issue in our `sanitize()` code. What's happening is that in code like this:
```
    return_trace (version.sanitize (c) &&
		  likely (version.major == 1) &&
		  varRecords.sanitize (c, this));
```
after inlining `version.sanitize (c)` and realizing that it doesn't change the value of `version.major`, the `version.major` load is moved earlier by the compiler to pipeline with the `version.sanitize(c)` code. The read then becomes illegal if sanitize is to fail...

This probably affects hundreds of our sanitize functions. Even the ones like:
```
  template <typename TLookup> 
  bool sanitize (hb_sanitize_context_t *c) const 
  { 
    TRACE_SANITIZE (this); 
    if (unlikely (!u.version.sanitize (c))) return_trace (false); 
    switch (u.version.major) { 
    case 1: return_trace (u.version1.sanitize<TLookup> (c)); 
#ifndef HB_NO_BEYOND_64K 
    case 2: return_trace (u.version2.sanitize<TLookup> (c)); 
#endif 
    default: return_trace (true); 
    } 
  }   
```
Or even simpler:
```
  bool sanitize (hb_sanitize_context_t *c) const 
  { 
    TRACE_SANITIZE (this); 
    return_trace (c->check_struct (this) && 
↦       ↦         extensionLookupType != T::SubTable::Extension); 
  } 
```

Basically any access to `*this` can possibly be moved before the `c->check_struct (this)` and cause a problem.

I don't think adding a barrier inside the sanitize machinery itself can fix this, just probably make it less likely to happen.

**Comment by behdad** [comment]
To properly fix this we need to audit all `sanitize()` methods and add read barriers between `check_struct` and subsequent reads. And similarly for the original case reported.

**Comment by behdad** [comment]
Fix in https://github.com/harfbuzz/harfbuzz/pull/4480

**Comment by behdad** [comment]
What I did was to introduce `hb_barrier ()` as a compiler read barrier, and then we have to add that every time we want to access a field after first sanitizing it (in the sanitize code itself). We also need to add it after any version check in the rest of the code.  See PR for 100s of examples.

**Comment by behdad** [comment]
Here's a sample in godbolt showing the problem:
![image](https://github.com/harfbuzz/harfbuzz/assets/142891/a1cdb2a2-9608-4fd3-8b2a-6cf498206a73)
Note how `version` is loaded even if the first check failed.

**Comment by behdad** [comment]
With a barrier:
![image](https://github.com/harfbuzz/harfbuzz/assets/142891/121a4395-518a-46c3-a533-30b55f51b6d5)
